cranelift-wasm/src/sections_translator.rs

```diff
@@ -351,3 +351,46 @@ pub fn parse_data_section<'data>(
 
     Ok(())
 }
+
+/// Parses the Name section of the wasm module.
+pub fn parse_name_section<'data>(
+    mut names: NameSectionReader<'data>,
+    environ: &mut dyn ModuleEnvironment<'data>,
+) -> WasmResult<()> {
+    while let Ok(subsection) = names.read() {
+        match subsection {
+            wasmparser::Name::Function(function_subsection) => {
+                if let Some(function_names) = function_subsection
+                    .get_map()
+                    .ok()
+                    .and_then(parse_function_name_subsection)
+                {
+                    for (index, name) in function_names {
+                        environ.declare_func_name(index, name)?;
+                    }
+                }
+            }
+            wasmparser::Name::Local(_) | wasmparser::Name::Module(_) => {}
+        };
+    }
+    Ok(())
+}
+
+fn parse_function_name_subsection<'data>(
+    mut naming_reader: NamingReader<'data>,
+) -> Option<HashMap<FuncIndex, &str>> {
+    let mut function_names = HashMap::new();
```

Comment:
(could be a SecondaryMap here too)

Reply:
Gave this a whirl, but it feels like `HashMap` _is_ what we need here? I found that because `SecondaryMap` does not expose an `insert` method, there isn't an elegant way to force an early return when a duplicate name is found. Would also be happy to ignore the possibility of duplicate values altogether here, but that seemed like a design decision I'd leave up to you :slightly_smiling_face: